Mathlib/Analysis/Normed/Group/SeparationQuotient.lean

```diff
@@ -0,0 +1,174 @@
+/-
+Copyright (c) 2024 Yoh Tanimoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yoh Tanimoto
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.Normed.Group.Hom
+
+/-!
+# Lifts of maps to separation quotients of seminormed groups
+
+For any `SeminormedAddCommGroup M`, a `NormedAddCommGroup` instance has been defined in
+`Mathlib.Analysis.Normed.Group.Uniform`.
+
+## Main definitions
+
+We use `M` and `N` to denote seminormed groups.
+All the following definitions are in the `NullSubgroup` namespace. Hence we can access
+`NullSubgroup.normedMk` as `normedMk`.
+
+* `normedAddCommGroupQuotient` : The normed group structure on the quotient by the null subgroup.
+    This is an instance so there is no need to explicitly use it.
+
+* `normedMk` : the normed group hom from `M` to `SeparationQuotient M`.
+
+* `lift f hf`: implements the universal property of `SeparationQuotient M`. Here
+    `(f : NormedAddGroupHom M N)`, `(hf : ∀ s ∈ nullSubgroup M, f s = 0)` and
+    `lift f hf : NormedAddGroupHom (SeparationQuotient M) N`.
+
+* `IsQuotient`: given `f : NormedAddGroupHom M N`, `IsQuotient f` means `N` is isomorphic
+    to a quotient of `M` by the null subgroup, with projection `f`. Technically it asserts `f` is
+    surjective and the norm of `f x` is the infimum of the norms of `x + m` for `m` in `f.ker`.
+
+## Main results
+
+* `norm_normedMk` : the operator norm of the projection is `1` if the subspace is not dense.
+
+* `IsQuotient.norm_lift`: Provided `f : normed_hom M N` satisfies `IsQuotient f`, for every
+     `n : N` and positive `ε`, there exists `m` such that `f m = n ∧ ‖m‖ < ‖n‖ + ε`.
+
+
+## Implementation details
+
+For any `SeminormedAddCommGroup M`, we define a norm on `SeparationQuotient M` by
+`‖x‖ = ‖mk x‖` using the lift.
+
+-/
+
+
+noncomputable section
+
+open SeparationQuotient Metric Set Topology NNReal
+
+variable {M N : Type*} [SeminormedAddCommGroup M] [SeminormedAddCommGroup N]
+
+namespace SeparationQuotientAddGroup
+
+-- lemma inseparable_iff_norm_zero (x y : M) : Inseparable x y ↔ ‖x - y‖ = 0 := by
+--   rw [Metric.inseparable_iff, dist_eq_norm]
+
+variable (x : SeparationQuotient M)
+
+variable (z : M)
+
+/-- If for `(m : M)` it holds that `mk m = 0`, then `m  ∈ nullSubgroup`. -/
```

Comment:
I don't think this docstring is adding anything.

Reply:
sorry, I corrected docstrings